File: go/2024/solutions/day03.go
package solutions

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

type Day03 struct{}

func (d Day03) Part1(raw string) {
	fmt.Println("Day 3 Part 1")
	r, _ := regexp.Compile(`mul\((\d+),(\d+)\)`)
	sum := 0
	for _, match := range r.FindAllSubmatch([]byte(raw), -1) {
		n1, _ := strconv.Atoi(string(match[1]))
		n2, _ := strconv.Atoi(string(match[2]))
		sum += n1 * n2
	}
	fmt.Println("Solution:", sum)
}

func (d Day03) Part2(raw string) {
	fmt.Println("Day 3 Part 2")

	doParts := make([]string, 0)
	for _, dos := range strings.Split(raw, "do()") {
		doParts = append(doParts, strings.Split(dos, "don't()")[0])
	}
	doInput := strings.Join(doParts, " ")

	r, _ := regexp.Compile(`mul\((\d+),(\d+)\)`)
	sum := 0
	for _, match := range r.FindAllSubmatch([]byte(doInput), -1) {
		n1, _ := strconv.Atoi(string(match[1]))
		n2, _ := strconv.Atoi(string(match[2]))
		sum += n1 * n2
	}
	fmt.Println("Solution:", sum)
}

File: rust/2024/src/solutions/day_03.rs
use itertools::Itertools;
use regex::Regex;

use super::day_trait::Day;

pub struct Day03;
impl Day for Day03 {
    fn day(&self) -> u8 {
        03
    }

    fn part_1(&self, raw: &str) {
        println!("Day {} part 1", self.day());

        let re = Regex::new("mul\\((\\d+),(\\d+)\\)").unwrap();
        let sum: usize = re
            .captures_iter(raw)
            .map(|c| {
                let (_, [a, b]) = c.extract();
                a.parse::<usize>().unwrap() * b.parse::<usize>().unwrap()
            })
            .sum();

        println!("Answer is {}", sum);
    }

    fn part_2(&self, raw: &str) {
        println!("Day {} part 2", self.day());

        let re = Regex::new("mul\\((\\d+),(\\d+)\\)").unwrap();
        let sum: usize = re
            .captures_iter(
                &raw.split("do()")
                    .map(|part| part.split("don't()").collect_vec()[0])
                    .collect_vec()
                    .join(" "),
            )
            .map(|c| {
                let (_, [a, b]) = c.extract();
                a.parse::<usize>().unwrap() * b.parse::<usize>().unwrap()
            })
            .sum();

        println!("Answer is {}", sum);
    }
}
